csharp/src/Ice/UdpEndpoint.cs

```diff
@@ -150,16 +141,94 @@ public override void IceWritePayload(OutputStream ostr)
             : base(communicator, protocol, host, port, sourceAddress)
         {
             McastInterface = mcastInterface;
-            McastTtl = mttl;
+            _mcastTtl = mttl;
             _connect = connect;
             HasCompressionFlag = compressionFlag;
         }
 
+        // Constructor for unmarshaling.
+        // TODO: should be ice1-only
         internal UdpEndpoint(InputStream istr, Communicator communicator, Protocol protocol)
             : base(istr, communicator, protocol)
         {
             _connect = false;
-            HasCompressionFlag = istr.ReadBool();
+            if (protocol == Protocol.Ice1)
+            {
+                HasCompressionFlag = istr.ReadBool();
+            }
+            else
+            {
+                SkipUnknownOptions(istr, istr.ReadSize());
+            }
+        }
+
+        // Constructor for URI parsing
+        internal UdpEndpoint(
+            Communicator communicator,
+            Protocol protocol,           // TODO: temporary, should always be ice1
+            string host,
+            ushort port,
+            Dictionary<string, string> options,
+            bool oaEndpoint,
+            string endpointString)
+            : base(communicator, protocol, host, port, options, oaEndpoint)
+        {
+            _connect = false; // TODO: what is this connect about?
```

Comment:
@bentoi do you remember why is the _connect required for UdpEndpoints?

Reply:
I don't really know, I'm pretty sure it's there since the beginning... if you look at the code, you'll see that it's used to connect the server-side UDP endpoint to the first client that sends a datagram. Perhaps it was to support some UDP hole punching use-case... as previously mentioned we should review the UDP implementation and decide what to keep and what to abandon (most likely this _connect attribute).